server/datastore/mysql/scim.go

```diff
@@ -0,0 +1,331 @@
+			return ctxerr.Wrap(ctx, err, "insert scim user")
+		}
+
+		id, err := result.LastInsertId()
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "insert scim user last insert id")
+		}
+		user.ID = uint(id) // nolint:gosec // dismiss G115
+		userID = user.ID
+
+		return insertEmails(ctx, tx, user)
+	})
+	return userID, err
+}
+
+// ScimUserByID retrieves a SCIM user by ID
+func (ds *Datastore) ScimUserByID(ctx context.Context, id uint) (*fleet.ScimUser, error) {
+	const query = `
+		SELECT
+			id, external_id, user_name, given_name, family_name, active
+		FROM scim_users
+		WHERE id = ?
+	`
+	user := &fleet.ScimUser{}
+	err := sqlx.GetContext(ctx, ds.reader(ctx), user, query, id)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, notFound("scim user").WithID(id)
+		}
+		return nil, ctxerr.Wrap(ctx, err, "select scim user")
+	}
+
+	// Get the user's emails
+	emails, err := ds.getScimUserEmails(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	user.Emails = emails
+
+	return user, nil
+}
+
+// ScimUserByUserName retrieves a SCIM user by username
+func (ds *Datastore) ScimUserByUserName(ctx context.Context, userName string) (*fleet.ScimUser, error) {
+	const query = `
+		SELECT
+			id, external_id, user_name, given_name, family_name, active
+		FROM scim_users
+		WHERE user_name = ?
+	`
+	user := &fleet.ScimUser{}
+	err := sqlx.GetContext(ctx, ds.reader(ctx), user, query, userName)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, notFound("scim user")
+		}
+		return nil, ctxerr.Wrap(ctx, err, "select scim user by userName")
+	}
+
+	// Get the user's emails
+	emails, err := ds.getScimUserEmails(ctx, user.ID)
+	if err != nil {
+		return nil, err
+	}
+	user.Emails = emails
+
+	return user, nil
+}
+
+// ReplaceScimUser replaces an existing SCIM user in the database
+func (ds *Datastore) ReplaceScimUser(ctx context.Context, user *fleet.ScimUser) error {
+	return ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+		// Update the SCIM user
+		const updateUserQuery = `
+		UPDATE scim_users SET
+			external_id = ?,
+			user_name = ?,
+			given_name = ?,
+			family_name = ?,
+			active = ?
+		WHERE id = ?`
+		result, err := tx.ExecContext(
+			ctx,
+			updateUserQuery,
+			user.ExternalID,
+			user.UserName,
+			user.GivenName,
+			user.FamilyName,
+			user.Active,
+			user.ID,
+		)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "update scim user")
+		}
+
+		rowsAffected, err := result.RowsAffected()
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "get rows affected for update scim user")
+		}
+		if rowsAffected == 0 {
+			return notFound("scim user").WithID(user.ID)
+		}
+
+		// We assume that email is not blank/null.
+		// However, we do not assume that email/type are unique for a user. To keep the code simple, we:
+		// 1. Delete all existing emails
+		// 2. Insert all new emails
+		// This is less efficient and can be optimized if we notice a load on these tables in production.
```

Comment:
Thanks for the callout. Based on troubles we've seen with batch updates to host profiles, for example, we should definitely keep a close eye on this and not hesitate to optimize if we start to see signs of poor performance. Hopefully the load here isn't too bad.